design/proof-erasure-coding.md

```diff
@@ -0,0 +1,244 @@
+Should we use this to encode a 1TB file, we'd end up with blocks of 16MB, far
+too large to be practical in a SNARK.
+
+Design space
+------------
+
+This limits the choices that we can make. The limiting factors seem to be:
+
+- Maximum number of blocks, determined by the field size of the erasure coding
+  algorithm
+- Number of blocks per proof, which determines how likely we are to detect
+  missing blocks
+- Capacity of the SNARK algorithm; how many bytes can we hash in a reasonable
+  time inside the SNARK
+
+From these limiting factors we can derive:
+
+- Block size
+- Maximum slot size; the maximum amount of data that can be verified with a
+  proof
+- Erasure coding memory requirements
+
+For example, when we use the leopard library, with a Galois field of 2^16, and
+require 80 blocks to be sampled per proof, and we can implement a SNARK that can
+hash 80*64K bytes, then we have:
+
+- Block size: 64KB
+- Maximum slot size: 4GB (2^16 * 64KB)
+- Erasure coding memory: > 128KB (2^16 * 16 bits)
+
+Which has the disadvantage of having a rather low maximum slot size of 4GB. When
+we want to improve on this to support e.g. 1TB slot sizes, we'll need to either
+increase the capacity of the SNARK, increase the field size of the erasure
+coding algorithm, or decrease the durability guarantees.
+
+> The [accompanying spreadsheet][4] allows you to explore the design space
+> yourself
+
+Increasing SNARK capacity
+-------------------------
+
+Increasing the computational capacity of SNARKs is an active field of study, but
+it is unlikely that we'll see an implementation of SNARKS that is 100-1000x
+faster before we launch Codex. Better hashing algorithms are also being designed
+for use in SNARKS, but it is equally unlikely that we'll see such a speedup here
+either.
+
+Decreasing durability guarantees
+--------------------------------
+
+We could reduce the durability guarantees by requiring e.g. 20 instead of 80
+blocks per proof. This would still give us a probability of detecting missing
+data of 1 - 0.5^20, which is 0.999999046, or "six nines". Arguably this is still
+good enough. Choosing 20 blocks per proof allows for slots up to 16GB:
+
+- Block size: 256KB
+- Maximum slot size: 16GB (2^16 * 256KB)
+- Erasure coding memory: > 128KB (2^16 * 16 bits)
+
+Erasure coding field size
+-------------------------
+
+If we could perform erasure coding on a field of around 2^20 to 2^30, then this
+would allow us to get to larger slots. For instance, with a field of at least
+size 2^24, we could support slot sizes up to 1TB:
+
+- Block size: 64KB
+- Maximum slot size: 1TB (2^24 * 64KB)
+- Erasure coding memory: > 48MB (2^24 * 24 bits)
+
+We are however unaware of any implementations of reed solomon that use a field
+size larger than 2^16 and still be efficient O(N log(N)). [FastECC][2] uses a
+prime field of 20 bits, but it lacks a decoder and a byte encoding scheme. The
+paper ["An Efficient (n,k) Information Dispersal Algorithm Based on Fermat
```

Comment:
The private repo available to you, https://github.com/Bulat-Ziganshin/private/tree/main/FastECC2 , includes decoder.

This code isn't fully systematic for binary data as it works in non-binary fields, but efficient encodings with very little waste exist. OTOH, Leopard is pretty similar (only replacing FFT with "additive FFT") and works in binary fields, so why do you need other code at all? Leopard should be extensible to GF(2^32), ask the author (or just provide GF(2^32) operations, in addition to 8 and 16 included in the repo).

Reply:
This is actually very interesting, we do care about non-binary fields because they might perform better in some for some specific cryptographic verification schemes, SNARKS, polynomial commitments, etc...

Reply:
Multiplication in GF(p) with large p is implemented as `a*b mod p` which for fixed p can be somewhat optimized with widely known "division by a known constant" algorithm that involves several usual (binary) multiplications.

OTOH, optimized multiplication in GF(2^n) is implemented by splitting numbers into 4/8-bit "digits" and using the multiplication table lookups for their multiplication.

If binary multiplications are faster than table lookups, then indeed GF(p) arithmetic can be faster than GF(2^n).

====

So, what do you need now?
1. algorithms (both fastecc and leopard) were published by 3rd-party researchers in 2014-2015, so they are public property.
2. I can quickly add public domain monicker to my fastecc2 code, and overall plan to push it into the public fastecc repo (sigh)
3. Efficient recoding to binary is possible when p is close to any 2^n, in particular check https://github.com/Bulat-Ziganshin/FastECC/blob/master/GF.md#efficient-data-packing
4. The long-standing approach to discovering RS codecs' potential performance should be to build their computation model, e.g. fastecc coder executes k*n*log2(n) multiplications in GF(p), which results in k2*n*log2(n) 32-bit multiplications (we need to find k and then k2). Then by learning how much time we spend executing GF(P) or 32-bit multiplication in a given SNARK setup, we will be able to predict the performance.

How else can I help?

Reply:
> [FastECC][2] uses a prime field of 20 bits

Check [lucky numbers](https://github.com/Bulat-Ziganshin/FastECC/blob/master/GF.md#lucky-number-choosing-the-best-base-for-computations). The prime I used for efficiency is 32-bit one, there is a nice 61-bit one, and we can extend the approach to 61*2 bits 🤣

Reply:
https://github.com/catid/leopard/issues/2 discusses implementation of >16 bits in Leopard

Reply:
> we do care about non-binary fields because they might perform better for some specific cryptographic verification schemes, SNARKS, polynomial commitments, etc...

at least in this documentation, SNARK is used only to hash a small subset of blocks, e.g. ECC codec itself has to work only on CPU directly

Reply:
Hey, super helpful stuff, but I (we) need some time to digest it :-)

Reply:
Overall, if we use GF(p^n) codec, we should stick to p values close to any 2^k, since they allow efficient recoding to/from binary data (like losing only 1 bit per 4 KiB with p=0xFFF00001). [This encoding scheme](https://github.com/Bulat-Ziganshin/FastECC/blob/master/GF.md#efficient-data-packing) is generic.

Other implications:
- if used p is close to 2^k, so we recode to k-bit numbers, RS coding will have to use blocks of `i*k` bits. E.g. with k=61, we can't use exactly 1 KiB source blocks, instead we have to use `8235=61*135`-bit blocks, adding 43 zero bits to each 1 KiB block, and then having to keep the extra 43 (non-zero!) bits for each 1KiB parity block. These extra data can be kept together with block checksum and other control data, so it's possible, but adds some complications.
- If p>2^k, then source data are fine (all k-bit numbers are <p), but parity data need to be recoded and extra bits stored in control data. If p<2^k, it's the opposite - we have to recode source data and then reliably store extra bits (e.g. one bit required to distinguish between 0 and 0xFF if p=0xFF) since we will need these extra bits to successfully restore this source block from other data if it was lost.

Reply:
Hi Bulat! Thanks for the great comments.

> FastECC decoder is included in the private repo available to you

Thank you, I didn't know that it was already implemented. I'll update the text to reflect this.

> This code isn't fully systematic for binary data as it works in non-binary fields

> 3\. Efficient recoding to binary is possible when p is close to any 2^n, in particular check https://github.com/Bulat-Ziganshin/FastECC/blob/master/GF.md#efficient-data-packing

If I understand this correctly, then this would mean that the parity data would be a  little bigger than normal, because of the need to represent elements from a field that is slightly larger than 2^n. But I expect that we can leave the original data untouched, because we can always encode it into field elements when needed? So in that aspect it is a systematic code, because it leaves the original data untouched?

> OTOH, Leopard employs almost the same algo (only replacing FFT with "additive FFT") and works in binary fields, so why do you need other code at all?

Good question, I was under the impression that this would lead to prohibitively large tables in Leopard, and that this wouldn't be the case with FastECC. But I could very well be mistaken.